### DISCUSSION TITLE
Uv editor (Polygon2D) improvements - bottom tab, uv wrap, offset, color editor

### DISCUSSION THREAD

**Discussion by aXu-AP** [godotengine/godot-proposals#4118] [Editor]
Hi, I have worked on a project which uses Polygon2Ds for both character (cut-out animation) and levels, and have noticed some things that could be more streamlined with the uv editor.
These could be filed as bugs as they are inconsistent from rest of Godot:
- **Zooming centers on origo instead of mouse** - this differs from how zooming is handled in scene editor. PR godotengine/godot#83731
- **Hotkeys** for move, rotate and scale don't exist.

These would be small improvements on workflow:
- **Set pivot point visually (offset)**. If you have made cutout animations, you know the clunkiness of getting offset point right by numerical values. It could be an option to just click where you want it. PR: godotengine/godot#83612
- **Show repeated texture**. I'm making levels with textured polygons and I'd want to adjust UV values to match brick seams. Check image below and try to figure it out. This feature should of course be linked to tiled option for texture.
![nimetön](https://user-images.githubusercontent.com/1621768/155785362-6408a8c4-d12f-4e4a-9eaa-97be795088a4.png)

And these ones I think would be a major booster to usability:
- **Move the editor to bottom panel**. I feel this is where it should belong, together with animation, tileset and other editors. This would allow for instant preview when setting uvs, polygons and bone weights. Separate proposal: #8209
- **Color editor**. Currently the only way to add vertex colors is by writing them by hand or setting from code. Tbh I haven't bothered with either. However the potential would be in handcrafting lighting in level, for example. Separate proposal: #8195

What is your take on these? Have you noticed other places where uv editor could be improved?

**Comment by insomniacUNDERSCORElemon** [comment]
- allow the UV editor (the non-UV parts that is) to be accessed without an assigned texture (because textureless polygons)
- 1-click (or even painting) for tris and quads, flip border tool
- a way to quick-generate faces that were not manually created (maybe treat that differently so modification is easier)

--------------------

Aside from a vertex painter I'd also like to see a bit more than that. An index system that allows for quick iteration or even animation, even better if a layer system could be had (including saving a layer as a resource) to make creation even easier. On a very basic level see https://github.com/godotengine/godot/issues/25120, at one point the vertex colors you did not define were inherited from the polygon's color (so it allowed *one* color that you could quickly change for many points).

Layers and indexed color will allow for quickly changing colors without shading being an issue. For instance, if you do shading on a balloon manually now it'd be tedious if you decide to change the color later... or if you want multiple colors of balloon.

![ArcoLinux_2020-12-14_14-22-16](https://user-images.githubusercontent.com/4397725/190884488-245312b9-c71a-4552-b6f8-1f6fc2e38500.png)

**Reply by aXu-AP** [reply]
> allow the UV editor (the non-UV parts that is) to be accessed without an assigned texture (because textureless polygons)

This I agree and it kind of happened as a byproduct when I tried moving the editor to bottom dock. However currently opening the UV editor creates uv array if it doesn't exist, which might not be what user wants, so some prompt needs to be added there.

> 1-click (or even painting) for tris and quads, flip border tool

I'm not sure what you are referring to, could you elaborate?

> a way to quick-generate faces that were not manually created (maybe treat that differently so modification is easier)

I think this could be a button on toolbar when `Polygons` edit mode is active. For simplicity, I think it should clear existing polygons and make new ones. As for algorithm for deciding how the vertices should connect I'm not sure. It should try to minimize differences between polygon areas. Probably this is a solved problem, or we could try to come up with a novel solution. Anyways, there seems to be some proposals about this: #1309 #3791

Sorry for raising these questions only now, but I'm now working on this so I thought I'd ask for clarifying.

**Reply by insomniacUNDERSCORElemon** [reply]
> I'm not sure what you are referring to, could you elaborate?

It's a problem when you add internal points. You need to click every vertex to make a face. (move mouse) Click. (move mouse) Click. (move mouse) Click. But if you select tri or quad beforehand, this should be possible with one click (or click and drag) based on the centerpoint and some highlighting to show what face you are about to create. And on top of auto-generation, flipping the edge orientation of 2 tris is a known thing easier than re-creating 2 faces (also _beautifying_, where _all_ edges are re-oriented a certain way).

> For simplicity, I think it should clear existing polygons and make new ones

My point with that was you might need faces in a specific orientation in just _one_ specific spot, that's why I say auto-generate the unfilled parts (and maybe even remember the non-auto part). In practice I don't know how much of an issue this'd be particularly with the better tools I've described for creating/editing faces. 

Though I think it does matter for both vertex colors and UV stretching even if it's not straightforward on what the difference will be.

--------------------

Also not sure if it's been discussed but I forgot 2 improvements that could be made to the _normal_ editor:

* when creating a new polygon: if the object's pivot is outside of the polygon created, place the pivot in the new center. Particularly useful with 1st creation since pivot defaults to 0,0
* the ability to select multiple vertices (and transform them)

**Reply by aXu-AP** [reply]
* Flipping edge could be maybe ctrl clicking the edge between 2 tris?
* I think auto-filling polys and then making necessary changes sounds more usual workflow? Might work the other way around too. Again, we need to come up with an actual algorithm for this one. Let's see if I can make a prototype some time.
* I don't know if moving pivot automatically is a good idea. Setting pivot to polygon's center could be a menu option, however.
* Edit tool should work as selection tool if dragged from empty space. The transformation tools are in place already, so this seems like a good idea.

**Reply by insomniacUNDERSCORElemon** [reply]
> I don't know if moving pivot automatically is a good idea

I thought about this and it's why I specified _outside_ of the created polygon. For instance if you were to create something intended as a character arm and thus containing 0,0 (or wherever the existing pivot is, even on an edge) the pivot wouldn't move.

I am not sure why someone would want a polygon in the middle of the screen with the pivot at a distance like 0,0. And this is annoying when creating multiple polygons and needing to also specify the pivot (moreso if you forgot and added child nodes, so you need to de-parent before you change the pivot). A center option would be good, but in this case needing to center every single polygon still isn't ideal.

-----------------------

EDIT: Also to be clear when I talk about pivot re-calculating I specifically mean re-creating the polygon in the main editor, not any operation in the sub-editor. At least that could be a distinction. 

On a similar note I could see using the sub-editor for fine-tuned details and thus having a different (likely smaller) grid size than the main scene editor.

**Comment by aXu-AP** [comment]
I have been tinkering with this a bit and I'm starting to believe that implementing some if not all of these is possible for me, along with misc. enhancements to the uv editor.
So far I have moved the editor to the bottom dock to allow live previewing uv edits:
![uv](https://github.com/godotengine/godot-proposals/assets/1621768/78aafbf1-c070-45de-a62a-2412a13bb604)

And added pivot editing:
![pivot](https://github.com/godotengine/godot-proposals/assets/1621768/097916c3-7f3d-4284-a891-4f4f8b0d38a9)

As such I think now would be good time to discuss if proposed things are what is actually wanted and how to proceed.

**Comment by aXu-AP** [comment]
Made a separate proposal for color editor: #8195

**Comment by aXu-AP** [comment]
A few more:
* Upon opening the editor should be centered to the texture, currently everything sits in top-left corner. PR: godotengine/godot#83731
* Reorder the modes to match workflow: Points, Polygons, UV, (Color), Bones. (you don't start with UV's, place some points first!)
* ~~Moving, resizing, rotating the polygon in Points mode don't commit changes to undo-redo.~~ EDIT: godotengine/godot#83657

Also an extra argument for moving the editor to bottom dock: undo doesn't work inside the editor. I guess this could be added to it, but moving the editor would give this for free.

**Reply by Calinou** [reply]
Centering the texture and reordering the modes as you suggested makes sense to me.

Does the undo-redo issue occur when moving individual points or editing their UVs? It's likely a bug and should be reported on the main Godot repository. There is an [old issue](https://github.com/godotengine/godot/issues/24853) mentioning something similar but it may not be the same thing.

**Reply by aXu-AP** [reply]
No, it occurs only when transforming all points. I opened an issue for it as it doesn't seem to be addressed in that old issue: godotengine/godot#83657

**Comment by aXu-AP** [comment]
Made a separate proposal for bottom dock editor: #8209

**Comment by aXu-AP** [comment]
> * **Hotkeys** for move, rotate and scale don't exist.

Elaborate on this a bit, as here's actually 2 issues:
* W, E and S hotkeys for Move, Rotate and Scale don't exist.
* For select tool transforming is available with dragging with modifier keys pressed - however these aren't the same as in canvas editor. I suggest to change them to match (change shift to alt).

**Comment by kitbdev** [comment]
It would be nice if there was a background square or outline so the 0-1 range is visible, for textures with transparency.
This can help the user notice if there are scaling issues, since there is currently no reference.

Ex. Blender's UV grid

![image](https://github.com/user-attachments/assets/8794f80d-a6f0-478d-a7e8-0adea6d92606)

**Reply by aXu-AP** [reply]
Could you elaborate what "scaling issues" you refer to?

**Reply by kitbdev** [reply]
This can help if the user's texture has transparency and is scaled too small.
Or if it does not have a texture, but you still want to configure the UVs for it. For instance, if you don't have a texture yet or the UVs need to be used for something procedurally and it shouldn't have a texture.

Personally, I noticed this when I was trying to generate a polygon through code (in the editor) and when I opened the UV editor, it looked fine, but I actually accidentally scaled them incorrectly.

**Reply by Calinou** [reply]
The code from https://github.com/godotengine/godot/pull/98716 can probably be copied over to the Polygon2D UV editor.

**Reply by aXu-AP** [reply]
I can see the benefit here, yeah.
For Polygon2D without texture however, this won't work, since the node scales the uvs according to texture size (ie. not `0-1` but `0-128` for 128 px texture). So a polygon without texture wouldn't know what size of a square to show.
